modules/rosapps/applications/devutils/lnktool/lnktool.cpp

```diff
@@ -0,0 +1,870 @@
+            break;
+        e = HRESULT_CODE(e); // "WIN32_FROM_HRESULT"
+    }
+    wprintf(Error < 0 ? L"Error 0x%.8X %s\n" : L"Error %d %s\n", Error, buf);
+    return Error;
+}
+
+static int SuccessOrReportError(int Error)
+{
+    return Error ? ErrMsg(Error) : Error;
+}
+
+static LONG StrToNum(PCWSTR in)
+{
+    PWCHAR end;
+    LONG v = wcstol(in, &end, 0);
+    if (v == LONG_MAX)
+        v = wcstoul(in, &end, 0);
+    return (end > in) ? v : 0;
+}
+
+static inline HRESULT Seek(IStream *pStrm, int Move, int Origin = FILE_CURRENT, ULARGE_INTEGER *pNewPos = NULL)
+{
+    LARGE_INTEGER Pos;
+    Pos.QuadPart = Move;
+    return pStrm->Seek(Pos, Origin, pNewPos);
+}
+
+static HRESULT IL_LoadFromStream(IStream *pStrm, PIDLIST_ABSOLUTE *ppidl)
+{
+    HMODULE hShell32 = LoadLibraryA("SHELL32");
+    FARPROC p = GetProcAddress(hShell32, (char*)26); // NT5
+    if (!p)
+        p = GetProcAddress(hShell32, (char*)846); // NT6
+    return ((int (WINAPI*)(IStream*,PIDLIST_ABSOLUTE*))p)(pStrm, ppidl);
+}
+
+static HRESULT SHParseName(PCWSTR Path, PIDLIST_ABSOLUTE *ppidl)
+{
+    HMODULE hShell32 = LoadLibraryA("SHELL32");
+    FARPROC p = GetProcAddress(hShell32, "SHParseDisplayName");
+    if (p)
+        return ((int (WINAPI*)(PCWSTR,void*,PIDLIST_ABSOLUTE*,UINT,UINT*))p)(Path, NULL, ppidl, 0, NULL);
+    return SHILCreateFromPath(Path, ppidl, NULL);
+}
+
+static HRESULT SHParseNameEx(PCWSTR Path, PIDLIST_ABSOLUTE *ppidl, bool Simple)
+{
+    if (!Simple)
+        return SHParseName(Path, ppidl);
+    *ppidl = SHSimpleIDListFromPath(Path);
+    return *ppidl ? S_OK : E_FAIL;
+}
+
+static INT_PTR SHGetPidlInfo(LPCITEMIDLIST pidl, SHFILEINFOW &shfi, UINT Flags, UINT Depth = 0)
+{
+    LPITEMIDLIST pidlDup = NULL, pidlChild;
+    if (Depth > 0)
+    {
+        if ((pidl = pidlDup = pidlChild = ILClone(pidl)) == NULL)
+            return 0;
+        while (Depth--)
+        {
+            if (LPITEMIDLIST pidlNext = ILGetNext(pidlChild))
+                pidlChild = pidlNext;
+        }
+        pidlChild->mkid.cb = 0;
+    }
+    INT_PTR ret = SHGetFileInfoW((PWSTR)pidl, 0, &shfi, sizeof(shfi), Flags | SHGFI_PIDL);
+    ILFree(pidlDup);
+    return ret;
+}
+
+static void PrintOffsetString(PCSTR Name, LPCVOID Base, UINT Min, UINT Ansi, UINT Unicode, PCSTR Indent = "")
+{
+    if (Unicode && Unicode >= Min)
+        wprintf(L"%hs%hs=%ls", Indent, Name, (PWSTR)((BYTE*)Base + Unicode));
+    else
+        wprintf(L"%hs%hs=%hs", Indent, Name, Ansi >= Min ? (char*)Base + Ansi : "");
+    wprintf(L"\n"); // Separate function call in case the (untrusted) input ends with a DEL character
+}
+
+static void Print(PCSTR Name, REFGUID Guid, PCSTR Indent = "")
+{
+    WCHAR Buffer[39];
+    StringFromGUID2(Guid, Buffer, _countof(Buffer));
+    wprintf(L"%hs%hs=%ls\n", Indent, Name, Buffer);
+}
+
+template<class V, class T>
+static void DumpFlags(V Value, T *pInfo, UINT Count, PCSTR Prefix = NULL)
+{
+    V Handled = 0;
```

Comment:
nit: If you remove the flag from `Value` (`&= ~flag`), then you don't need this `Handled`